board/freescale/mx6ullevk/mx6ullevk.c

```diff
@@ -276,12 +285,291 @@ int board_late_init(void)
+struct mx6_ddr_sysinfo eh8010_ddr_sysinfo = {
+	.dsize = 0,
+	.cs_density = 20,
+	.ncs = 1,
+	.cs1_mirror = 0,
+	.rtt_wr = 2,
+	.rtt_nom = 1,		/* RTT_Nom = RZQ/2 */
+	.walat = 0,		/* Write additional latency */
+	.ralat = 5,		/* Read additional latency */
+	.mif3_mode = 3,		/* Command prediction working mode */
+	.bi_on = 1,		/* Bank interleaving enabled */
+	.sde_to_rst = 0x10,	/* 14 cycles, 200us (JEDEC default) */
+	.rst_to_cke = 0x23,	/* 33 cycles, 500us (JEDEC default) */
+	.ddr_type = DDR_TYPE_DDR3,
+	.refsel = 0,	/* Refresh cycles at 64KHz */
+	.refr = 1,	/* 2 refresh commands per refresh cycle */
+};
+
+/* IS43TR16512B-107MBLI 1GB
+ * https://www.issi.com/WW/pdf/43-46TR16512B-81024BL.pdf */
+struct mx6_ddr_sysinfo eh8020_ddr_sysinfo = {
+	.dsize = 0,
+	.cs_density = 24,
+	.ncs = 1,
+	.cs1_mirror = 0,
+	.rtt_wr = 2,
+	.rtt_nom = 1,		/* RTT_Nom = RZQ/2 */
+	.walat = 0,		/* Write additional latency */
+	.ralat = 5,		/* Read additional latency */
+	.mif3_mode = 3,		/* Command prediction working mode */
+	.bi_on = 1,		/* ?: Bank interleaving enabled */
+	.sde_to_rst = 0x10,	/* ?: 14 cycles, 200us (JEDEC default) */
+	.rst_to_cke = 0x23,	/* RST_to_CKE: 33 cycles, 500us (JEDEC default) */
+	.ddr_type = DDR_TYPE_DDR3,
+	.refsel = 0,		/* REF_SEL: Refresh cycles at 64KHz */
+	.refr = 1,		/* REFR: 2 refresh commands per refresh cycle */
+};
+
+/* K4B4G1646E-BMMA 500 MB
+ * https://semiconductor.samsung.com/resources/data-sheet/DS_K4B4G1646E_BY_M_Rev1_11-0.pdf */
+static struct mx6_ddr3_cfg eh8010_mem_ddr = {
+	.mem_speed = 800,
+	.density = 4,
+	.width = 16,
+	.banks = 8,
+	.rowaddr = 15,
+	.coladdr = 10,
+	.pagesz = 2,
+	.trcd = 1375,
+	.trcmin = 4875,
+	.trasmin = 3500,
+};
+
+/* IS43TR16512B-107MBLI 1GB
+ * https://www.issi.com/WW/pdf/43-46TR16512B-81024BL.pdf */
+static struct mx6_ddr3_cfg eh8020_mem_ddr = {
+	.mem_speed = 1600,
+	.density = 8,
+	.width = 16,
+	.banks = 8,
+	.rowaddr = 16,
+	.coladdr = 10,
+	.pagesz = 2,
+	.trcd = 1375,
+	.trcmin = 4875,
+	.trasmin = 3500,
+};
+
+
+static void setup_iomux_uart(void)
+{
+	imx_iomux_v3_setup_multiple_pads(uart1_pads, ARRAY_SIZE(uart1_pads));
+}
+
+static void spl_dram_init(void)
+{
+	struct mx6_ddr3_cfg		*mem_ddr;
+	struct mx6ul_iomux_ddr_regs	*mx6_ddr_ioregs;
+	struct mx6ul_iomux_grp_regs	*mx6_grp_ioregs;
+	struct mx6_ddr_sysinfo		*ddr_sysinfo;
+	struct mx6_mmdc_calibration	*mx6_mmcd_calib;
+
+	unsigned int siklu_board_id = get_siklu_board_id();
+
+	switch (siklu_board_id) {
+		case SKL_BOARD_TYPE_PCB195:
+			puts("SKL_BOARD_TYPE_PCB195\n");
+		case SKL_BOARD_TYPE_PCB213:
+			puts("SKL_BOARD_TYPE_PCB213\n");
+		case SKL_BOARD_TYPE_PCB217:
+			puts("SKL_BOARD_TYPE_PCB217\n");
+			/* K4B4G1646E-BMMA 500 MB
+			 * https://semiconductor.samsung.com/resources/data-sheet/DS_K4B4G1646E_BY_M_Rev1_11-0.pdf */
+			mem_ddr		= &eh8010_mem_ddr;
+			mx6_ddr_ioregs	= &eh8010_ddr_ioregs;
+			mx6_grp_ioregs	= &eh8010_grp_ioregs;
+			ddr_sysinfo	= &eh8010_ddr_sysinfo;
+			mx6_mmcd_calib	= &eh8010_mmcd_calib;
```

Comment:
It looks like `mem_ddr`, `mx6_ddr_ioregs`, and `mx6_grp_ioregs` are the same for both boards. So I think they should share the same structs, with their name indicating that they are shared, maybe something like `eh80x0_mem_ddr`.

Reply:
Not clear. What does "are the same for both boards" mean?

Reply:
> Not clear. What does "are the same for both boards" mean?

The fields in `eh8010_ddr_ioregs` are initialized exactly the same as in `eh8020_ddr_ioregs`. We can have a single `eh80x0_ddr_ioregs` with these values, and use it for `mx6ul_dram_iocfg()` directly.

Reply:
It is likely that I will change these initializations when I do the fine tuning. They are likely to diverge.

Reply:
Splitting the struct when there is a need is easy. Removing duplication is hard.

Reply:
I tried to remove this duplication and ran into runtime problems that I do not have time to debug. IMHO removing this duplication is a temporary commit, sing the code is likely to diverge again later.